packages/server/src/time-control.ts

```diff
@@ -40,100 +54,326 @@ export function ValidateTimeControlBase(
+    }
+
+    switch (game.config.time_control.type) {
+      case TimeControlType.Absolute:
+      case TimeControlType.Fischer: {
+        const times: ITimeControlBase = game.time_control;
+
+        if (times === undefined || times.forPlayer === undefined) {
+          // old game with no moves
+          return null;
+        }
+
+        const playerTime = times.forPlayer[playerNr];
+
+        if (playerTime === undefined || playerTime.onThePlaySince === null) {
+          // player has not played a move yet
+          // or player is not on the play
+          return null;
+        }
+
+        const timeoutTime =
+          playerTime.onThePlaySince.getTime() + playerTime.remainingTimeMS;
+
+        return timeoutTime - new Date().getTime();
+      }
+      default: {
         console.error(`game with id ${game.id} has invalid time control type`);
         return;
+      }
     }
   }
 }
 
 class TimeHandlerParallelMoves implements ITimeHandler {
+  private _timeoutService: TimeoutService;
+  constructor() {
+    this._timeoutService = getTimeoutService();
+  }
+
+  initialState(
+    variant: string,
+    config: IConfigWithTimeControl,
+  ): TimeControlParallel {
+    const numPlayers = makeGameObject(variant, config).numPlayers();
+
+    const timeControl: TimeControlParallel = {
+      moveTimestamps: [],
+      forPlayer: {},
+    };
+
+    for (let i = 0; i < numPlayers; i++) {
+      timeControl.forPlayer[i] = {
+        remainingTimeMS: config.time_control.mainTimeMS,
+        onThePlaySince: null,
+        stagedMoveAt: null,
+      };
+    }
+
+    switch (config.time_control.type) {
+      case TimeControlType.Absolute: {
+        // nothing to do
+        break;
+      }
+
+      case TimeControlType.Fischer: {
+        // nothing to do
+        break;
+      }
+
+      default: {
+        console.error("received config with invalid time control type");
+      }
+    }
+
+    return timeControl;
+  }
+
   handleMove(
-    _game: GameResponse,
-    _game_obj: AbstractGame<unknown, unknown>,
-    _playerNr: number,
-    _move: string,
-  ): Promise<void> {
-    console.log(
-      "time control handler for parallel moves is not implemented yet",
-    );
-    return;
+    game: GameResponse,
+    game_obj: AbstractGame<unknown, unknown>,
+    playerNr: number,
+  ): ITimeControlBase {
+    const config = game.config as IConfigWithTimeControl;
+
+    switch (config.time_control.type) {
+      case TimeControlType.Absolute: {
+        let timeControl = game.time_control as TimeControlParallel;
+        // could happen for old games
+        if (timeControl === undefined) {
+          timeControl = this.initialState(game.variant, config);
+        }
+
+        const timestamp = new Date();
+
+        timeControl.moveTimestamps.push(timestamp);
+        timeControl.forPlayer[playerNr].stagedMoveAt = timestamp;
+        this._timeoutService.clearPlayerTimeout(game.id, playerNr);
+
+        // check if the round finishes with this move
```

Comment:
I think that a round() method could be useful. As for the different time control handler, I believe we could unify the handlers if we use the same staged-moves logic from the parallel variant on sequencial moves variants. There a player would "stage a move", and because there is only one player relevant to this round, the round immediately proceeds. However, I believe that the current implementation for parallel moves might not handle such a scenario correctly -- for the reason see my comment in time-control.ts lines 196-198